orchestra/static/orchestra/task/js/required_fields.js

```diff
@@ -9,45 +9,73 @@
    */
   serviceModule.factory('requiredFields', function($rootScope, orchestraService) {
     var requiredFields = {
-      fields: [],
+      fields: {},
       invalid: [],
+      validators: {
+        'input-checkbox': [
+          function(elem) {
+            return elem.checked;
+          }
+        ],
+        'input-text': [
+          function(elem) {
+            return elem.value && elem.value.length > 0;
+          }
+        ],
+      },
       setup: function(data) {
         /**
          * Sets up the base data on which to validate fields.
          */
         this.data = data;
       },
-      require: function(fields) {
+      require: function(fieldType, field) {
         /**
-         * Sets a field as required. Fields are specified by
-         * dot-delimited key strings (e.g., `key0.key1.key2`).
+         * Sets a field as required. Fields are HTML elements to be
+         * checked by one or more validators according to their field
+         * type.
          */
-        this.fields = this.fields.concat(fields)
+        if (Object.keys(this.fields).indexOf(fieldType) >= 0) {
+          this.fields[fieldType].push(field);
+        }
+        else {
+          this.fields[fieldType] = [field];
+        }
       },
       validate: function() {
         /**
-         * Validates required fields. For the required field
-         * `key0.key1.key2` to be valid, `this.data[key0][key1][key2]`
-         * must exist and not be falsy.
+         * Validates required fields according to their registered
+         * validators.
          */
         var requiredFields = this;
         requiredFields.invalid = [];
-        requiredFields.fields.forEach(function(field) {
-          // For each field, check that each successive key exists and
-          // is not falsy.
-          var obj = requiredFields.data;
-          var keys = field.split('.');
-          for (var i in keys) {
-            var key = keys[i];
-            obj = obj[key];
-            if (!obj) {
-              requiredFields.invalid.push(field)
-              break;
+        for (var fieldType in requiredFields.fields) {
+          var validators = requiredFields.validators[fieldType];
+          var fields = requiredFields.fields[fieldType];
+          fields.forEach(function(field) {
+            var success = true;
+            validators.forEach(function(validator) {
+              success = success && validator(field);
+            })
```

Comment:
should we catch the case where someone requires validation for a non-existent field type? e.g., fields contains `'input-radio'` but we don't support it? Here, validators will be undefined and `forEach ` will barf confusingly, I think.